autopaho/auto.go

```diff
@@ -326,3 +331,30 @@ func (c *ConnectionManager) Publish(ctx context.Context, p *paho.Publish) (*paho
 	}
 	return cli.Publish(ctx, p)
 }
+
+// UseRouter executes a function that is passed the client's Router
+// We pass a function to UseRouter instead of returning a reference to cli.Router to ensure we lock and unlock
+// the mutex protecting it's access
+func (c *ConnectionManager) UseRouter(fn func(paho.Router) error) error {
+	c.mu.Lock()
+	cli := c.cli
+	c.mu.Unlock()
```

Comment:
No, the other functions in the autopaho.ClientManager followed this pattern and they needed to synchronise access to the client too, so I followed that. It works since a copy by value occurs assigning a new instance of the paho.Client to the local variable cli, then releasing the lock protecting the client. This local instance of the client will then be used and cleaned up by the GC once the function returns.